From 7994c2faf40ff8cc3145045b9cb23cc0c4c09f00 Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Mon, 28 Jul 2025 16:34:40 +0530
Subject: [PATCH] WLANRTSYS-92027 validate acs flag during channel switch after
 radar detect and reset the second best channel param .

Issue : acs_init_done flag gets cleared after second best channel CAC
started.
Fix : Use conf_acs flag to validate ACS is enable or disable and reset
the second best channel param in interface disable place.
---
 src/ap/drv_callbacks.c       |  4 ----
 src/ap/hostapd.c             |  2 ++
 src/ap/mxl_acs_proprietary.c | 29 +----------------------------
 src/ap/mxl_dfs_common.c      | 10 +++-------
 src/ap/mxl_drv_callbacks.c   | 15 ---------------
 src/ap/mxl_drv_callbacks.h   |  1 -
 src/ap/mxl_hostapd.c         |  8 ++++++++
 src/ap/mxl_zwdfs.c           |  2 +-
 8 files changed, 15 insertions(+), 56 deletions(-)

diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 44613f6db..66b3bc366 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2188,10 +2188,6 @@ static void hostapd_event_dfs_cac_finished(struct hostapd_data *hapd,
 	hostapd_dfs_complete_cac(hapd->iface, 1, radar->freq, radar->ht_enabled,
 				 radar->chan_offset, radar->chan_width,
 				 radar->cf1, radar->cf2);
-#ifdef CONFIG_VENDOR_MXL
-	/* csa_iface reset if we do CAC manually at run time. So check csa pointer to do CAC with second best channel */
-	mxl_check_csa_and_get_second_best_channel(hapd);
-#endif /* CONFIG_VENDOR_MXL */
 }
 
 
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index aff3d219e..51b20d626 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4743,6 +4743,8 @@ int hostapd_disable_iface(struct hostapd_iface *hapd_iface)
 	mxl_acs_unset_bgscan(hapd_iface);
 	/* stop existing bgscan_interval timer */
 	mxl_acs_stop_bgscan_timer(hapd_iface->bss[0]);
+	/* reset second best channel params */
+	mxl_acs_clear_second_bg_cac_param(hapd_iface);
 #endif /* CONFIG_VENDOR_MXL */
 
 	/* same as hostapd_interface_deinit without deinitializing ctrl-iface */
diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 7e9046952..0a64454de 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -2089,11 +2089,6 @@ void add_channel_pair_candidates(struct hostapd_iface *iface, channel_pair *chan
 			    (acs_is_40_mhz_overlap_chan(&channel_pairs[i])))
 				continue;
 
-			if (is_excluded_opclass_chan(iface, channel_pairs[i].primary, 40) ||
-				is_excluded_opclass_chan(iface, channel_pairs[i].secondary, 40)) {
-				continue;
-			}
-
 			if (channel_pairs[i].primary < channel_pairs[i].secondary)
 				candidates[iface->mxl_iface.num_candidates].freq = chan->freq;
 			else
@@ -2260,8 +2255,7 @@ void mxl_acs_init_candidate_table(struct hostapd_iface *iface)
 
 		if (!acs_unii4_chan_disabled(iface, chan) &&
 		    (iface->conf->mxl_conf.acs_use24overlapped || (chan->chan == 1) ||
-		    (chan->chan == 6) || (chan->chan == 11) || (chan->freq >= 5000)) &&
-		    !is_excluded_opclass_chan(iface, chan->chan, 20)) {
+		    (chan->chan == 6) || (chan->chan == 11) || (chan->freq >= 5000))) {
 			if (chan->flag & HOSTAPD_CHAN_RADAR)
 				candidates[iface->mxl_iface.num_candidates].radar_affected = true;
 
@@ -3561,28 +3555,9 @@ int mxl_acs_recalc_ranks_and_set_chan (struct hostapd_iface *iface, int switch_r
 
 		wpa_printf(MSG_DEBUG, "ACS: recalculate candidate idx %d", i);
 
-		if (iface->mxl_iface.sec_best_ch.flag) {
-			acs_print_cand_no_file(fp, iface, i);
-			acs_cand_skip = true;
-		}
-
 		acs_print_info_header(MSG_DEBUG);
 		acs_print_cand_no(MSG_DEBUG, iface, i);
 
-		if (iface->mxl_iface.sec_best_ch.ch_switch_done) {
-			first_rank_freq = iface->mxl_iface.sec_best_ch.freq;
-			first_rank_width = iface->mxl_iface.sec_best_ch.width;
-			iface->mxl_iface.sec_best_ch.ch_switch_done = false;
-			wpa_printf(MSG_DEBUG, "ACS: ch_switch_done sec_best_ch freq : %d sec_best_ch width : %d", iface->mxl_iface.sec_best_ch.freq, iface->mxl_iface.sec_best_ch.width);
-		}
-
-		if (iface->mxl_iface.sec_best_ch.first_rank_idx && iface->mxl_iface.sec_best_ch.flag) {
-			if (channels_overlap(first_rank_freq, first_rank_width, candidates[i].freq, candidates[i].width)) {
-				wpa_printf(MSG_ERROR, "channel overlapped  first freq %d, : candidates[i].freq %d", candidates[iface->mxl_iface.sec_best_ch.first_rank_idx].freq, candidates[i].freq);
-				continue;
-			}
-		}
-
 		candidates[i].exclude_reason = acs_to_exclude_candidate(iface, i, &now);
 
 		if (iface->mxl_iface.sec_best_ch.flag) {
@@ -4968,8 +4943,6 @@ int mxl_acs_check_bgscan(struct hostapd_iface *iface)
 	if (iface->state == HAPD_IFACE_ENABLED)
 		iface->mxl_iface.acs_reenable_iface = 1;
 
-	iface->conf->mxl_conf.acs_init_done = 0;
-
 	return mxl_acs_start(iface);
 }
 
diff --git a/src/ap/mxl_dfs_common.c b/src/ap/mxl_dfs_common.c
index 9f7cd9cff..f1c6045fc 100644
--- a/src/ap/mxl_dfs_common.c
+++ b/src/ap/mxl_dfs_common.c
@@ -618,12 +618,6 @@ void mxl_dfs_restore_original_cw(struct hostapd_iface *iface)
 bool mxl_validate_second_best_chan(struct hostapd_iface *iface)
 {
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-#ifdef CONFIG_ACS
-	if(!iface->conf->mxl_conf.acs_init_done) {
-		wpa_printf(MSG_DEBUG, "acs_init_done is not set");
-		return false;
-	}
-#endif /* CONFIG_ACS */
 	if (iface->mxl_iface.sec_best_ch.bg_cac) {
 		wpa_printf(MSG_DEBUG, "Second best channel bg_cac set");
 		return true;
@@ -638,8 +632,10 @@ bool mxl_switch_second_best_channel(struct hostapd_iface *iface)
 {
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
 #ifdef CONFIG_ACS
-	if(!iface->conf->mxl_conf.acs_init_done)
+	if(!iface->conf->acs) {
+		wpa_printf(MSG_DEBUG, "%s ZWDFS: acs is disabled after zwdfs CAC on second best channel",  __func__);
 		return false;
+	}
 #endif /* CONFIG_ACS */
 	if (iface->mxl_iface.sec_best_ch.bg_cac && (iface->dfs_domain == HOSTAPD_DFS_REGION_FCC))
 	{
diff --git a/src/ap/mxl_drv_callbacks.c b/src/ap/mxl_drv_callbacks.c
index 5eed7661f..67d19f71d 100644
--- a/src/ap/mxl_drv_callbacks.c
+++ b/src/ap/mxl_drv_callbacks.c
@@ -248,18 +248,3 @@ void mxl_check_ch_switch_status_and_get_second_best_channel(int finished, struct
 #endif
 	return;
 }
-
-void mxl_check_csa_and_get_second_best_channel(struct hostapd_data *hapd)
-{
-#ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-	if(!hapd->iface->conf->mxl_conf.acs_init_done) {
-		return;
-	}
-	if (hapd->iface->mxl_iface.sec_best_ch.flag) {
-		if (hapd->iface->mxl_iface.zwdfs_info.csa_iface == NULL) {
-			mxl_get_second_best_channel(hapd->iface);
-		}
-	}
-#endif
-	return;
-}
diff --git a/src/ap/mxl_drv_callbacks.h b/src/ap/mxl_drv_callbacks.h
index d27fdbd03..ed4cb9af0 100644
--- a/src/ap/mxl_drv_callbacks.h
+++ b/src/ap/mxl_drv_callbacks.h
@@ -27,5 +27,4 @@ void mxl_hostapd_event_eapol_rx(struct hostapd_data *hapd, const u8 *src,
 				int link_id,
 				u16 vlan_id);
 void mxl_check_ch_switch_status_and_get_second_best_channel(int finished, struct hostapd_data *hapd);
-void mxl_check_csa_and_get_second_best_channel(struct hostapd_data *hapd);
 #endif /* MXL_DRV_CALLBACKS_H */
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 9b8024ac5..303655fc1 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -3672,3 +3672,11 @@ void mxl_dynamic_wmm_param_init_wrap(struct hostapd_data *hapd) {
 #endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
 	return;
 }
+
+void mxl_acs_clear_second_bg_cac_param(struct hostapd_iface *iface)
+{
+#ifdef CONFIG_VENDOR_MXL_PROPRIETARY
+	memset(&iface->mxl_iface.sec_best_ch, 0, sizeof(iface->mxl_iface.sec_best_ch));
+#endif /* CONFIG_VENDOR_MXL_PROPRIETARY */
+	return;
+}
diff --git a/src/ap/mxl_zwdfs.c b/src/ap/mxl_zwdfs.c
index e37d07b7d..519ee8c18 100644
--- a/src/ap/mxl_zwdfs.c
+++ b/src/ap/mxl_zwdfs.c
@@ -719,7 +719,7 @@ fail:
 
 int mxl_hostapd_handle_zwdfs_switch(struct hostapd_iface *iface, int success, int freq, int chan_width, int cf1, int cf2)
 {
-	wpa_printf(MSG_DEBUG, "ZWDFS: Switching channel triggered for channel: %d", iface->conf->channel);
+	wpa_printf(MSG_DEBUG, "%s ZWDFS: Switching channel is triggered", __func__);
 	if (mxl_hostapd_is_zwdfs_iface(iface)) {
 		if (mxl_dfs_use_radar_background(iface)) {
 			if (mxl_hostapd_dfs_is_background_event(iface, freq, cf1)) {
-- 
2.43.0

